internal/protoc/downloader.go

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package protoc
+
+import (
+	"archive/zip"
+	"bytes"
+	"crypto/sha512"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+
+	"github.com/uber/prototool/internal/x/settings"
+	"github.com/uber/prototool/internal/x/vars"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+type downloader struct {
+	logger    *zap.Logger
+	cachePath string
+	protocURL string
+	config    settings.Config
+
+	lock sync.RWMutex
+	// the looked-up and verified to exist base path
+	cachedBasePath string
+}
+
+func newDownloader(config settings.Config, options ...DownloaderOption) *downloader {
+	downloader := &downloader{
+		config: config,
```

Comment:
It looks like the `config` is only used to access the `ProtobufVersion`. Let's consider just having the version as an attribute here.

Reply:
Either/or is fine. I figured in principle I might want other things, and to just pass the config, but doesn't matter.

Reply:
We can easily flatten those fields in the `downloader` as they are needed. I doubt we'll need more than the version here. It's a small detail, but it would make the definition slightly simpler in my opinion.